halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/deployment/v1/DeploymentCRDGenerator.java

```diff
@@ -44,96 +45,125 @@
 @Slf4j
 public class DeploymentCRDGenerator {
 
-  @Autowired ServiceProviderFactory serviceProviderFactory;
+  @Autowired KubectlServiceProvider kubectlServiceProvider;
 
-  @Autowired Yaml yamlParser;
+  @Autowired HalconfigDirectoryStructure halconfigDirectoryStructure;
 
-  @Autowired StrictObjectMapper objectMapper;
+  @Autowired List<SpinnakerServiceGenerator> generators;
 
-  @Autowired HalconfigDirectoryStructure halconfigDirectoryStructure;
+  @Autowired Yaml yamlParser;
 
-  public String generateCR(
-      DeploymentConfiguration deploymentConfiguration, String storeType, String storeName) {
-    log.info("Parsing Halyard config");
-    KubectlServiceProvider serviceProvider =
-        (KubectlServiceProvider) serviceProviderFactory.create(deploymentConfiguration);
+  @Autowired ObjectMapper objectMapper;
 
+  public String generateCR(
+      DeploymentConfiguration deploymentConfiguration, String serviceName, String apiGroupVersion) {
     log.info("Resolving configuration");
-    CRConfig crConfig = new CRConfig();
+    CRConfig crConfig =
+        new CRConfig().setDeploymentConfiguration(deploymentConfiguration).setName(serviceName);
+
+    SpinnakerServiceGenerator generator = getGenerator(apiGroupVersion);
 
     // Runtime settings are needed when generating profiles and we generate profile to gather
     SpinnakerRuntimeSettings runtimeSettings =
-        serviceProvider.buildRuntimeSettings(deploymentConfiguration);
+        kubectlServiceProvider.buildRuntimeSettings(deploymentConfiguration);
+
+    log.info("Collecting service settings per service");
+    collectServiceSettings(runtimeSettings, crConfig);
 
-    // Gather all the profile file names
     Path userProfilePath =
         halconfigDirectoryStructure.getUserProfilePath(deploymentConfiguration.getName());
-    List<String> userProfileNames =
-        aggregateProfilesInPath(userProfilePath.toString(), "", crConfig);
 
-    log.info("Collecting required files per service");
-    serviceProvider.getServices().stream()
+    // Collect all existing profile files
+    List<String> userProfileNames = aggregateProfilesInPath(userProfilePath.toString(), "");
+
+    log.info("Collecting profile files per service");
+    // For each enable service, collect main profiles and required files
+    kubectlServiceProvider.getServices().stream()
```

Comment:
Is this only processing kubernetes configs? What about all other providers?

Reply:
`KubectlServiceProvider` holds the services to deploy via `Distributed`. It has nothing to do with the provider.